sync-core/src/main/java/com/cloudant/sync/datastore/AttachmentStreamFactory.java

```diff
@@ -0,0 +1,164 @@
+ * appropriate input/output stream chain to handle this.
+ *
+ * The input stream returned will allow reading unencrypted, unzipped data from
+ * disk.
+ *
+ * The output stream returned should have unencrypted, unzipped data written to
+ * its write method.
+ *
+ */
+class AttachmentStreamFactory {
+
+    /**
+     * Byte array if there is a valid AES key, null if attachments
+     * should not be encrypted.
+     */
+    private final byte[] key;
+
+    /**
+     * Creates a factory Attachment objects can use to read/write attachments.
+     *
+     * @param keyProvider Datastore's key provider object. This object can cope with key providers
+     *                    that return null keys.
+     */
+    public AttachmentStreamFactory(KeyProvider keyProvider) {
+        EncryptionKey key = keyProvider.getEncryptionKey();
+        if (key != null) {
+            this.key = key.getKey(); // guaranteed 256-bits from EncryptionKey
+        } else {
+            this.key = null;
+        }
+    }
+
+    /**
+     * Return a stream to be used to read from the file on disk.
+     *
+     * Stream's bytes will be unzipped, unencrypted attachment content.
+     *
+     * @param file File object to read from.
+     * @param encoding Encoding of attachment.
+     * @return Stream for reading attachment data.
+     * @throws IOException
+     */
+    public InputStream getInputStream(File file, Attachment.Encoding encoding) throws IOException {
+
+        InputStream is = new FileInputStream(file);
+
+        if (key != null) {
+            try {
+                is = new EncryptedAttachmentInputStream(is, key);
+            } catch (InvalidKeyException ex) {
+                // Replace with an IOException as we validate the key when opening
+                // the databases and the key should be the same -- it's therefore
+                // not worth forcing the developer to catch something they can't
+                // fix during file read; generic IOException works better.
+                throw new IOException("Bad key used to open file; check encryption key.", ex);
+            }
+        }
+
+        switch (encoding) {
+            case Plain:
+                break;  // nothing to do
+            case Gzip:
+                is = new GZIPInputStream(is);
+        }
+
+        return is;
+    }
+
+    /**
+     * Get stream for writing attachment data to disk.
+     *
+     * Opens the output stream using {@see FileUtils#openOutputStream(File)}.
+     *
+     * Data should be written to the stream un-encoded, un-encrypted.
+     *
+     * @param file File to write to.
+     * @param encoding Encoding to use.
+     * @return Stream for writing.
+     * @throws IOException
+     */
+    public OutputStream getOutputStream(File file, Attachment.Encoding encoding) throws
+            IOException {
+
+        OutputStream os = FileUtils.openOutputStream(file);
+
+        if (key != null) {
```

Comment:
It took me a moment to understand the logic here, can you summarise in a comment? My understanding is:
- If key is not null, wrap the OutputStream with a EncryptedAttachmentOutputStream to encrypt it
- If encoding is not Plain, wrap the OutputStream from the first step in a GZIPOutputStream